src/hotspot/os/windows/os_windows.cpp

```diff
@@ -1279,8 +1279,33 @@ void  os::dll_unload(void *lib) {
   }
 }
 
+static size_t format_message_fixup(char* buf, size_t end) {
+  if (end > 3) {
+    if (buf[end - 1] == '\n') end--;
+    if (buf[end - 1] == '\r') end--;
+    if (buf[end - 1] == '.')  end--;
+    buf[end] = '\0';
+  }
+  return end;
+}
+
 void* os::dll_lookup(void *lib, const char *name) {
-  return (void*)::GetProcAddress((HMODULE)lib, name);
+  void* ret = ::GetProcAddress((HMODULE)lib, name);
+  DWORD errval;
+  if ((errval = GetLastError()) != 0) {
```

Comment:
`GetProcAddress` doesn't clear last-error on success, so you should only be calling `GetLastError` if the return value was null.

Reply:
That said, why are you duplicating the logic from `os::lasterror` instead of just passing it the buf to fill in?
```
if (ret == nullptr) {
  char buf[512];
  os::lasterror(buf, sizeof(buf));
  log_debug(os)(...);
}
```

Reply:
> `GetProcAddress` doesn't clear last-error on success, so you should only be calling `GetLastError` if the return value was null.

It seem like you are correct. I didn't find anything on GetProcAddress manpage, but on GetLastError page it says: "If the function is not documented to set the last-error code, the value returned by this function is simply the most recent last-error code"

I'll update.

> That said, why are you duplicating the logic from `os::lasterror` instead of just passing it the buf to fill in?
> 
> ```
> if (ret == nullptr) {
>   char buf[512];
>   os::lasterror(buf, sizeof(buf));
>   log_debug(os)(...);
> }
> ```

It's not clear to me that GetProcAddress will not return NULL without setting GetLastError.
Thus calling os::lasterror could accidentally return error string for errno instead.
Maybe this worry have no basis, what you think?

Reply:
It should be easy to test what happens when you ask for a symbol known to not exist.

Reply:
Yes, but unfortunately I have no win32 env. nor access to anything other than GA for windows.
But I was thinking of the case where a Symbol have NULL as it's address.

From some old MSDN article:
"The only required array is the Export Address Table (EAT), which is an array of function pointers that contain the address of an exported function."

AFAICT GetProcAddress just returns this "pointer" and you may have NULL pointers in this array.

Hence I think the same issue with symbol exists but with NULL address as on posix is possible but not documented.

Reply:
From an API perspective that seems completely broken to me as it is impossible to distinguish the two cases and a call to GetLastError may return an unrelated error. But I couldn't find anything that indicated NULL could be a valid returned address. I found this description of the EAT and IIUC it will only have entries for exported functions:
https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#export-address-table

Reply:
Ok, updated win32.